cl/phase1/forkchoice/on_attestation.go

```diff
@@ -81,56 +85,174 @@ func (f *ForkChoiceStore) OnAttestation(attestation *solid.Attestation, fromBloc
+		return nil
+	}
+
+	// [IGNORE] The block being voted for (aggregate.data.beacon_block_root) has been seen (via both gossip and non-gossip sources) (a client MAY queue aggregates for processing once block is retrieved).
+	if _, ok := f.forkGraph.GetHeader(aggregateData.BeaconBlockRoot()); !ok {
 		return nil
 	}
 
-	activeIndiciesLength := uint64(len(targetState.shuffledSet))
+	// [REJECT] The committee index is within the expected range -- i.e. index < get_committee_count_per_slot(state, aggregate.data.target.epoch).
+	committeeCountPerSlot := headState.CommitteeCount(target.Epoch())
+	if aggregateData.CommitteeIndex() >= committeeCountPerSlot {
+		return fmt.Errorf("invalid committee index in aggregate and proof")
+	}
+	// [REJECT] The aggregate attestation's epoch matches its target -- i.e. aggregate.data.target.epoch == compute_epoch_at_slot(aggregate.data.slot)
+	if aggregateData.Target().Epoch() != f.computeEpochAtSlot(slot) {
+		return fmt.Errorf("invalid target epoch in aggregate and proof")
+	}
+	committee, err := headState.GetBeaconCommitee(slot, committeeIndex)
+	if err != nil {
+		return err
+	}
 
-	count := targetState.committeeCount(epoch, activeIndiciesLength) * f.beaconCfg.SlotsPerEpoch
-	start := (activeIndiciesLength * committeeIndex) / count
-	end := (activeIndiciesLength * (committeeIndex + 1)) / count
-	committeeLength := end - start
-	if !state.IsAggregator(f.beaconCfg, committeeLength, slot, committeeIndex, selectionProof) {
-		log.Warn("invalid aggregate and proof")
+	// [REJECT] aggregate_and_proof.selection_proof selects the validator as an aggregator for the slot -- i.e. is_aggregator(state, aggregate.data.slot, index, aggregate_and_proof.selection_proof) returns True.
+	if !state.IsAggregator(f.beaconCfg, uint64(len(committee)), committeeIndex, selectionProof) {
+		log.Warn("receveived aggregate and proof from invalid aggregator")
 		return fmt.Errorf("invalid aggregate and proof")
 	}
-	return f.OnAttestation(aggregateAndProof.Message.Aggregate, false, false)
+	// [REJECT] The aggregator's validator index is within the committee -- i.e. aggregate_and_proof.aggregator_index in get_beacon_committee(state, aggregate.data.slot, index).
+	if !slices.Contains(committee, aggregateAndProof.Message.AggregatorIndex) {
+		return fmt.Errorf("committee index not in committee")
+	}
+
+	if err := f.verifySignaturesOnAggregate(headState, aggregateAndProof); err != nil {
+		return err
+	}
+
+	// [REJECT] The aggregate attestation's target block is an ancestor of the block named in the LMD vote -- i.e. get_checkpoint_block(store, aggregate.data.beacon_block_root, aggregate.data.target.epoch) == aggregate.data.target.root
+	if f.Ancestor(aggregateData.BeaconBlockRoot(), f.computeStartSlotAtEpoch(target.Epoch())) != target.BlockRoot() {
+		return fmt.Errorf("invalid target block")
+	}
+
+	// Add to some sort of aggregation layer TODO(Giulio2002/domiwei)
```

Comment:
so, here the next step is to store `Aggregate *solid.Attestation` in `SignedAggregateAndProof.Message.Aggregate`, right? just want to align our thinking.
and will try merge aggregation if no bits overlapping, otherwise simply store them.

Reply:
when do we need to fetch aggregation result? should be queried by slot and committeeId?